Last updated: June 29th 2016

# Climate data exploration: a journey through Pandas

Welcome to a demo of Python's data analysis package called `Pandas`. Our goal is to learn about Data Analysis and transformation using Pandas while exploring datasets used to analyze climate change. 

## The story

The global goal of this demo is to provide the tools to be able to try and reproduce some of the analysis done in the IPCC global climate reports published in the last decade (see for example https://www.ipcc.ch/pdf/assessment-report/ar5/syr/SYR_AR5_FINAL_full.pdf). 

We are first going to load a few public datasets containing information about global temperature, global and local sea level infomation, and global concentration of greenhouse gases like CO2, to see if there are correlations and how the trends are to evolve, assuming no fundamental change in the system. For all these datasets, we will download them, visualize them, clean them, search through them, merge them, resample them, transform them and summarize them.

In the process, we will learn about:

Part 1:

    1. Loading data
    2. Pandas datastructures
    3. Cleaning and formatting data
    4. Basic visualization
   
Part 2:

    5. Accessing data
    6. Working with dates and times
    7. Transforming datasets
    8. Statistical analysis
    9. Data agregation and summarization
    10. Correlations and regressions
    11. Predictions from auto regression models

## Some initial setup

In [1]:
%matplotlib inline
import pandas as pd

import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_rows", 16)

LARGE_FIGSIZE = (12, 8)

In [2]:
# Change this cell to the demo location on YOUR machine
# %cd ~/Projects/pandas_tutorial/climate_timeseries/
%cd ~/Dropbox/Projects/pandas_tutorial/climate_timeseries
%ls

/home/ali/Dropbox/Projects/pandas_tutorial/climate_timeseries
climate_timeseries-Part1.ipynb  climate_timeseries-Part2.ipynb  data/  sandbox/


## 1. Loading data

More details, see http://pandas.pydata.org/pandas-docs/stable/io.html

To find all reading functions in pandas, ask ipython's tab completion:

In [3]:
# pd.read_table

In [4]:
pd.read_table?

### From a local text file

Let's first load some temperature data which covers all lattitudes. Since read_table is supposed to do the job for a text file, let's just try it:

In [5]:
filename = "data/temperatures/annual.land_ocean.90S.90N.df_1901-2000mean.dat"
full_globe_temp = pd.read_table(filename)
full_globe_temp

1880   -0.1591
0    1881   -0.0789
1    1882   -0.1313
2    1883   -0.1675
3    1884   -0.2485
4    1885   -0.2042
5    1886   -0.1695
6    1887   -0.2505
7    1888   -0.1605
..              ...
122  2003    0.5818
123  2004    0.5416
124  2005    0.6154
125  2006    0.5601
126  2007    0.5472
127  2008    0.4804
128  2009    0.5551
129  2010 -999.0000

[130 rows x 1 columns]

There is only 1 column! Let's try again stating that values are separated by any number of spaces:

In [6]:
full_globe_temp = pd.read_table(filename, sep="\s+")
full_globe_temp

1880   -0.1591
0    1881   -0.0789
1    1882   -0.1313
2    1883   -0.1675
3    1884   -0.2485
4    1885   -0.2042
5    1886   -0.1695
6    1887   -0.2505
7    1888   -0.1605
..    ...       ...
122  2003    0.5818
123  2004    0.5416
124  2005    0.6154
125  2006    0.5601
126  2007    0.5472
127  2008    0.4804
128  2009    0.5551
129  2010 -999.0000

[130 rows x 2 columns]

There are columns but the column names are 1880 and -0.1591!

In [7]:
full_globe_temp = pd.read_table(filename, sep="\s+", names=["year", "mean temp"])
full_globe_temp

year  mean temp
0    1880    -0.1591
1    1881    -0.0789
2    1882    -0.1313
3    1883    -0.1675
4    1884    -0.2485
5    1885    -0.2042
6    1886    -0.1695
7    1887    -0.2505
..    ...        ...
123  2003     0.5818
124  2004     0.5416
125  2005     0.6154
126  2006     0.5601
127  2007     0.5472
128  2008     0.4804
129  2009     0.5551
130  2010  -999.0000

[131 rows x 2 columns]

Since we only have 2 columns, one of which would be nicer to access the data (the year of the record), let's try using the `index_col` option:

In [8]:
full_globe_temp = pd.read_table(filename, sep="\s+", names=["year", "mean temp"], 
                                index_col=0)
full_globe_temp

mean temp
year           
1880    -0.1591
1881    -0.0789
1882    -0.1313
1883    -0.1675
1884    -0.2485
1885    -0.2042
1886    -0.1695
1887    -0.2505
...         ...
2003     0.5818
2004     0.5416
2005     0.6154
2006     0.5601
2007     0.5472
2008     0.4804
2009     0.5551
2010  -999.0000

[131 rows x 1 columns]

Last step: the index is made of dates. Let's make that explicit:

In [9]:
full_globe_temp = pd.read_table(filename, sep="\s+", names=["year", "mean temp"], 
                                index_col=0, parse_dates=True)
full_globe_temp

mean temp
year                          
1880-01-01 00:00:00    -0.1591
1881-01-01 00:00:00    -0.0789
1882-01-01 00:00:00    -0.1313
1883-01-01 00:00:00    -0.1675
1884-01-01 00:00:00    -0.2485
1885-01-01 00:00:00    -0.2042
1886-01-01 00:00:00    -0.1695
1887-01-01 00:00:00    -0.2505
...                        ...
2003-01-01              0.5818
2004-01-01              0.5416
2005-01-01              0.6154
2006-01-01              0.5601
2007-01-01              0.5472
2008-01-01              0.4804
2009-01-01              0.5551
2010-01-01           -999.0000

[131 rows x 1 columns]

### From a chunked file

Since every dataset can contain mistakes, let's load a different file with temperature data. NASA's GISS dataset is written in chunks: look at it in `data/temperatures/GLB.Ts+dSST.txt`

In [10]:
giss_temp = pd.read_table("data/temperatures/GLB.Ts+dSST.txt", sep="\s+", skiprows=7,
                          skip_footer=11, engine="python")
giss_temp

Year  Jan  Feb  Mar  Apr   May   Jun   Jul   Aug   Sep   Oct   Nov   Dec  \
0    1880  -34  -27  -22  -30   -16   -24   -19   -12   -20   -19   -16   -21   
1    1881  -13  -16   -2   -3    -3   -27   -12    -8   -18   -23   -28   -18   
2    1882    3    4   -2  -24   -20   -32   -27   -11   -11   -25   -25   -37   
3    1883  -38  -38  -12  -20   -20    -8    -3   -13   -19   -19   -28   -21   
4    1884  -20  -14  -31  -36   -33   -36   -31   -24   -29   -25   -29   -25   
5    1885  -57  -29  -19  -36   -35   -40   -28   -24   -17   -14   -14     0   
6    1886  -37  -43  -34  -23   -21   -30   -13   -21   -12   -22   -29   -18   
7    1887  -60  -43  -26  -34   -28   -25   -19   -28   -24   -33   -29   -40   
..    ...  ...  ...  ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   
135  2009   56   48   49   57    59    62    66    61    64    58    72    58   
136  2010   66   75   87   82    71    60    56    59    55    65    74    44   
137  2011   45   44   57   60    47    54    70    69    52    60    50    48   
138  2012   38   43   52   62    71    59    50    56    68    73    69    46   
139  2013   62   52   60   48    56    61    53    61    73    61    75    61   
140  2014   68   44   71   72    79    62    50    74    81    78    64    74   
141  2015   75   80   84   71  ****  ****  ****  ****  ****  ****  ****  ****   
142  Year  Jan  Feb  Mar  Apr   May   Jun   Jul   Aug   Sep   Oct   Nov   Dec   

      J-D  D-N   DJF   MAM   JJA   SON Year.1  
0     -22  ***  ****   -23   -18   -18   1880  
1     -14  -14   -17    -3   -15   -23   1881  
2     -17  -16    -4   -15   -23   -20   1882  
3     -20  -21   -38   -18    -8   -22   1883  
4     -28  -28   -18   -33   -31   -28   1884  
5     -26  -28   -37   -30   -31   -15   1885  
6     -25  -24   -26   -26   -21   -21   1886  
7     -32  -30   -40   -29   -24   -28   1887  
..    ...  ...   ...   ...   ...   ...    ...  
135    59   59    52    55    63    65   2009  
136    66   67    66    80    58    65   2010  
137    55   55    45    55    64    54   2011  
138    57   57    43    62    55    70   2012  
139    60   59    53    55    58    70   2013  
140    68   67    58    74    62    74   2014  
141  ****  ***    76  ****  ****  ****   2015  
142   J-D  D-N   DJF   MAM   JJA   SON   Year  

[143 rows x 20 columns]

**QUIZ:** What happens if you remove the `skiprows`? `skipfooter`? `engine`?

**EXERCISE:** Load some readings of CO2 concentrations in the atmosphere from the `data/greenhouse_gaz/co2_mm_global.txt` data file.

In [11]:
# Your code here
co2_concentration = pd.read_table("data/greenhouse_gaz/co2_mm_global.txt", sep="\s+", 
                                  parse_dates = [[0, 1]])
# parse two comlums(1 and 2) together inside a column
co2_concentration

year_month   decimal  average   trend
0   1980-01-01  1980.042   338.45  337.82
1   1980-02-01  1980.125   339.17  338.12
2   1980-03-01  1980.208   339.49  338.14
3   1980-04-01  1980.292   339.87  338.25
4   1980-05-01  1980.375   340.30  338.78
5   1980-06-01  1980.458   339.86  339.08
6   1980-07-01  1980.542   338.32  339.17
7   1980-08-01  1980.625   337.11  339.37
..         ...       ...      ...     ...
416 2014-09-01  2014.708   394.90  397.61
417 2014-10-01  2014.792   396.18  397.63
418 2014-11-01  2014.875   397.69  397.90
419 2014-12-01  2014.958   398.62  398.08
420 2015-01-01  2015.042   399.31  398.29
421 2015-02-01  2015.125   400.01  398.59
422 2015-03-01  2015.208   400.59  398.96
423 2015-04-01  2015.292   401.24  399.41

[424 rows x 4 columns]

### From a remote text file

So far, we have only loaded temperature datasets. Climate change also affects the sea levels on the globe. Let's load some datasets with the sea levels. The university of colorado posts updated timeseries for mean sea level globably, per 
hemisphere, or even per ocean, sea, ... Let's download the global one, and the ones for the northern and southern hemisphere.

That will also illustrate that to load text files that are online, there is no more work than replacing the filepath by a URL n `read_table`:

In [12]:
# Local backup: data/sea_levels/sl_nh.txt
northern_sea_level = pd.read_table("http://sealevel.colorado.edu/files/current/sl_nh.txt", 
                                   sep="\s+")
northern_sea_level

year  msl_ib(mm)  #version_2016_rel4
0    1992.9323      13.717                 NaN
1    1992.9595       1.328                 NaN
2    1992.9866     -13.375                 NaN
3    1993.0138     -24.723                 NaN
4    1993.0409     -29.231                 NaN
5    1993.0681     -32.538                 NaN
6    1993.0952     -35.588                 NaN
7    1993.1223     -26.579                 NaN
..         ...         ...                 ...
861  2016.3611      56.282                 NaN
862  2016.3883      63.273                 NaN
863  2016.4154      69.371                 NaN
864  2016.4426      75.595                 NaN
865  2016.4697      80.087                 NaN
866  2016.4969      81.619                 NaN
867  2016.5240      83.580                 NaN
868  2016.5512      87.190                 NaN

[869 rows x 3 columns]

In [13]:
# Local backup: data/sea_levels/sl_sh.txt
southern_sea_level = pd.read_table("http://sealevel.colorado.edu/files/current/sl_sh.txt", 
                                   sep="\s+")
southern_sea_level

year  msl_ib(mm)  #version_2016_rel4
0    1992.9323       2.517                 NaN
1    1992.9595      -7.645                 NaN
2    1992.9866      -2.267                 NaN
3    1993.0138       0.855                 NaN
4    1993.0409      -2.020                 NaN
5    1993.0681      -0.398                 NaN
6    1993.0952       4.102                 NaN
7    1993.1223       6.787                 NaN
..         ...         ...                 ...
861  2016.3611      66.331                 NaN
862  2016.3883      64.067                 NaN
863  2016.4154      68.305                 NaN
864  2016.4426      67.053                 NaN
865  2016.4697      65.238                 NaN
866  2016.4969      54.950                 NaN
867  2016.5240      53.514                 NaN
868  2016.5512      55.389                 NaN

[869 rows x 3 columns]

In [14]:
# The 2015 version of the global dataset:
# Local backup: data/sea_levels/sl_ns_global.txt
url = "http://sealevel.colorado.edu/files/2015_rel2/sl_ns_global.txt"
global_sea_level = pd.read_table(url, sep="\s+")
global_sea_level

year  msl_ib_ns(mm)  #version_2015_rel2
0    1992.9595         -5.818                 NaN
1    1992.9866         -7.525                 NaN
2    1993.0138         -9.215                 NaN
3    1993.0409        -11.796                 NaN
4    1993.0681        -11.291                 NaN
5    1993.0952         -9.569                 NaN
6    1993.1223         -3.714                 NaN
7    1993.1495         -2.471                 NaN
..         ...            ...                 ...
772  2014.9494         70.723                 NaN
773  2014.9766         70.522                 NaN
774  2015.0037         66.662                 NaN
775  2015.0309         64.804                 NaN
776  2015.0580         62.115                 NaN
777  2015.0852         68.589                 NaN
778  2015.1123         69.745                 NaN
779  2015.1395         75.007                 NaN

[780 rows x 3 columns]

There are clearly lots of cleanup to be done on these datasets. See below...

### From a local or remote HTML file

To be able to grab more local data about mean sea levels, we can download and extract data about mean sea level stations around the world from the PSMSL (http://www.psmsl.org/). Again to download and parse all tables in a webpage, just give `read_html` the URL to parse:

In [15]:
# Needs `lxml`, `beautifulSoup4` and `html5lib` python packages
# Local backup in data/sea_levels/Obtaining Tide Gauge Data.html
table_list = pd.read_html("http://www.psmsl.org/data/obtaining/")

In [16]:
table_list

[                 Station Name    ID    Lat.     Lon.  GLOSS ID   Country    \
 0                    HELIGMAN   231  60.200   19.300         NaN       ALA   
 1                KOBBAKLINTAR    63  60.033   19.883         NaN       ALA   
 2                    LEMSTROM    84  60.100   20.017         NaN       ALA   
 3             FOGLO / DEGERBY   249  60.032   20.385         NaN       ALA   
 4                   PAGO PAGO   539 -14.280 -170.690       144.0       ASM   
 5             BAHIA ESPERANZA   988 -63.300  -56.917       185.0       ATA   
 6           ARGENTINE ISLANDS   913 -65.246  -64.257       188.0       ATA   
 7            PUERTO SOBERANIA  1603 -62.483  -59.633       189.0       ATA   
 ...                       ...   ...     ...      ...         ...       ...   
 1462                   DANANG  1475  16.100  108.217         NaN       VNM   
 1463                   HONNGU  1003  18.800  105.767         NaN       VNM   
 1464                   HONDAU   841  20.667  106.80

In [17]:
# there is 1 table on that page which contains metadata about the stations where 
# sea levels are recorded
local_sea_level_stations = table_list[0]
local_sea_level_stations

Station Name    ID    Lat.     Lon.  GLOSS ID   Country    \
0                    HELIGMAN   231  60.200   19.300         NaN       ALA   
1                KOBBAKLINTAR    63  60.033   19.883         NaN       ALA   
2                    LEMSTROM    84  60.100   20.017         NaN       ALA   
3             FOGLO / DEGERBY   249  60.032   20.385         NaN       ALA   
4                   PAGO PAGO   539 -14.280 -170.690       144.0       ASM   
5             BAHIA ESPERANZA   988 -63.300  -56.917       185.0       ATA   
6           ARGENTINE ISLANDS   913 -65.246  -64.257       188.0       ATA   
7            PUERTO SOBERANIA  1603 -62.483  -59.633       189.0       ATA   
...                       ...   ...     ...      ...         ...       ...   
1462                   DANANG  1475  16.100  108.217         NaN       VNM   
1463                   HONNGU  1003  18.800  105.767         NaN       VNM   
1464                   HONDAU   841  20.667  106.800         NaN       VNM   
1465         CHARLOTTE AMALIE  1393  18.335  -64.920         NaN       VIR   
1466  LIME TREE BAY, ST CROIX  1447  17.693  -64.753         NaN       VIR   
1467    CHRISTIANSTED HARBOUR  2118  17.750  -64.705         NaN       VIR   
1468             LAMESHUR BAY  2119  18.317  -64.723         NaN       VIR   
1469                     ADEN    44  12.788   44.974         3.0       YEM   

            Date  Coastline    Station    
0     01/01/1980           60        251  
1     01/01/1980           60        261  
2     01/01/1980           60        271  
3     06/10/2015           60        281  
4     08/03/2016          745          1  
5     11/05/1999          999          1  
6     21/05/2015          999          3  
7     30/09/2004          999          5  
...          ...          ...        ...  
1462  17/11/2015          605         51  
1463  17/11/2015          605         61  
1464  17/11/2015          605         81  
1465  08/03/2016          939          1  
1466  08/03/2016          939         11  
1467  09/03/2016          939         15  
1468  09/03/2016          939         21  
1469  01/02/2016          485          1  

[1470 rows x 9 columns]

That table can be used to search for a station in a region of the world we choose, extract an ID for it and download the corresponding time series with the URL http://www.psmsl.org/data/obtaining/met.monthly.data/< ID >.metdata

## 2. Pandas DataStructures

For more details, see http://pandas.pydata.org/pandas-docs/stable/dsintro.html

Now that we have used `read_**` functions to load datasets, we need to understand better what kind of objects we got from them to learn to work with them.

### DataFrame, the pandas 2D structure

In [18]:
# Type of the object?
type(giss_temp)

pandas.core.frame.DataFrame

In [19]:
# Internal nature of the object
print(giss_temp.shape)
print(giss_temp.dtypes)

(143, 20)
Year      object
Jan       object
Feb       object
Mar       object
Apr       object
May       object
Jun       object
Jul       object
           ...  
Dec       object
J-D       object
D-N       object
DJF       object
MAM       object
JJA       object
SON       object
Year.1    object
dtype: object


Descriptors for the vertical axis (axis=0)

In [20]:
giss_temp.index

RangeIndex(start=0, stop=143, step=1)

Descriptors for the horizontal axis (axis=1)

In [21]:
giss_temp.columns

Index([u'Year', u'Jan', u'Feb', u'Mar', u'Apr', u'May', u'Jun', u'Jul', u'Aug',
       u'Sep', u'Oct', u'Nov', u'Dec', u'J-D', u'D-N', u'DJF', u'MAM', u'JJA',
       u'SON', u'Year.1'],
      dtype='object')

A lot of information at once including memory usage:

In [22]:
giss_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 143 entries, 0 to 142
Data columns (total 20 columns):
Year      143 non-null object
Jan       143 non-null object
Feb       143 non-null object
Mar       143 non-null object
Apr       143 non-null object
May       143 non-null object
Jun       143 non-null object
Jul       143 non-null object
Aug       143 non-null object
Sep       143 non-null object
Oct       143 non-null object
Nov       143 non-null object
Dec       143 non-null object
J-D       143 non-null object
D-N       143 non-null object
DJF       143 non-null object
MAM       143 non-null object
JJA       143 non-null object
SON       143 non-null object
Year.1    143 non-null object
dtypes: object(20)
memory usage: 22.4+ KB


### Series, the pandas 1D structure

A series can be constructed with the `pd.Series` constructor (passing a list or array of values) or from a `DataFrame`, by extracting one of its columns.

In [27]:
# Do we already have a series for the full_globe_temp?
type(full_globe_temp)

pandas.core.series.Series

In [30]:
full_globe_temp

year
1880-01-01 00:00:00     -0.1591
1881-01-01 00:00:00     -0.0789
1882-01-01 00:00:00     -0.1313
1883-01-01 00:00:00     -0.1675
1884-01-01 00:00:00     -0.2485
1885-01-01 00:00:00     -0.2042
1886-01-01 00:00:00     -0.1695
1887-01-01 00:00:00     -0.2505
                         ...   
2003-01-01               0.5818
2004-01-01               0.5416
2005-01-01               0.6154
2006-01-01               0.5601
2007-01-01               0.5472
2008-01-01               0.4804
2009-01-01               0.5551
2010-01-01            -999.0000
Name: mean temp, dtype: float64

In [28]:
# Since there is only one column of values, we can make this a Series without 
# loosing information:
full_globe_temp = full_globe_temp["mean temp"]

KeyError: 'mean temp'

Core attributes/information:

In [ ]:
print(type(full_globe_temp))
print(full_globe_temp.dtype)
print(full_globe_temp.shape)
print(full_globe_temp.nbytes)

Probably the most important attribute of a `Series` or `DataFrame` is its `index` since we will use that to, well, index into the structures to access te information:

In [ ]:
full_globe_temp.index

### NumPy arrays as backend of Pandas

It is always possible to fall back to a good old NumPy array to pass on to scientific libraries that need them: SciPy, scikit-learn, ...

In [ ]:
full_globe_temp.values

In [ ]:
type(full_globe_temp.values)

### Creating new DataFrames manually

`DataFrame`s can also be created manually, by grouping several Series together. Let's make a new frame from the 3 sea level datasets we downloaded above. They will be displayed along the same index. Wait, does that makes sense to do that?

In [ ]:
# Are they aligned?
southern_sea_level.year == northern_sea_level.year

In [ ]:
# So, are they aligned?
np.all(southern_sea_level.year == northern_sea_level.year)

So the northern hemisphere and southern hemisphere datasets are aligned. What about the global one?

In [ ]:
len(global_sea_level.year) == len(northern_sea_level.year)

For now, let's just build a DataFrame with the 2 hemisphere datasets then. We will come back to add the global one later...

In [ ]:
mean_sea_level = pd.DataFrame({"northern_hem": northern_sea_level["msl_ib(mm)"], 
                               "southern_hem": southern_sea_level["msl_ib(mm)"], 
                               "date": northern_sea_level.year})
mean_sea_level

Note: there are other ways to create DataFrames manually, for example from a 2D numpy array.

There is still the date in a regular column and a numerical index that is not that meaningful. We can specify the `index` of a `DataFrame` at creation. Let's try:

In [ ]:
mean_sea_level = pd.DataFrame({"northern_hem": northern_sea_level["msl_ib(mm)"], 
                               "southern_hem": southern_sea_level["msl_ib(mm)"]},
                               index = northern_sea_level.year)
mean_sea_level

Now the fact that it is failing show that Pandas does auto-alignment of values: for each value of the index, it searches for a value in each Series that maps the same value. Since these series have a dumb numerical index, no values are found. 

Since we know that the order of the values match the index we chose, we can replace the Series by their values only at creation of the DataFrame:

In [ ]:
mean_sea_level = pd.DataFrame({"northern_hem": northern_sea_level["msl_ib(mm)"].values, 
                               "southern_hem": southern_sea_level["msl_ib(mm)"].values},
                               index = northern_sea_level.year)
mean_sea_level

## 3. Cleaning and formatting data

The datasets that we obtain straight from the reading functions are pretty raw. A lot of pre-processing can be done during data read but we haven't used all the power of the reading functions. Let's learn to do a lot of cleaning and formatting of the data.

The GISS temperature dataset has a lot of issues too: useless numerical index, redundant columns, useless rows, placeholder (`****`) for missing values, and wrong type for the columns. Let's fix all this:

### Renaming columns

In [ ]:
# The columns of the local_sea_level_stations aren't clean: they contain spaces and dots.
local_sea_level_stations.columns

In [ ]:
# Let's clean them up a bit:
local_sea_level_stations.columns = [name.strip().replace(".", "") 
                                    for name in local_sea_level_stations.columns]
local_sea_level_stations.columns

We can also rename an index by setting its name. For example, the index of the `mean_sea_level` dataFrame could be called `date` since it contains more than just the year:

In [ ]:
mean_sea_level.index.name = "date"
mean_sea_level

### Setting missing values

In the full globe dataset, -999.00 was used to indicate that there was no value for that year. Let's search for all these values and replace them with the missing value that Pandas understand: `np.nan` 

In [ ]:
full_globe_temp == -999.000

In [ ]:
full_globe_temp[full_globe_temp == -999.000] = np.nan
full_globe_temp.tail()

### Choosing what is the index

In [ ]:
# We didn't set a column number of the index of giss_temp, we can do that afterwards:
giss_temp = giss_temp.set_index("Year")
giss_temp.head()

### Dropping rows and columns

In [ ]:
# 1 column is redundant with the index: 
giss_temp.columns

In [ ]:
# Let's drop it:
giss_temp = giss_temp.drop("Year.1", axis=1)
giss_temp

In [ ]:
# We can also just select the columns we want to keep:
giss_temp = giss_temp[[u'Jan', u'Feb', u'Mar', u'Apr', u'May', u'Jun', u'Jul', 
                       u'Aug', u'Sep', u'Oct', u'Nov', u'Dec']]
giss_temp

In [ ]:
# Let's remove all these extra column names (Year  Jan ...). They all correspond to the index "Year"
giss_temp = giss_temp.drop("Year")
giss_temp

Let's also set `****` to a real missing value (`np.nan`). We can often do it using a boolean mask, but that may trigger pandas warning. Another way to assign based on a boolean condition is to use the `where` method:

In [ ]:
#giss_temp[giss_temp == "****"] = np.nan
giss_temp = giss_temp.where(giss_temp != "****", np.nan)

In [ ]:
giss_temp.tail()

### Adding columns

While building the `mean_sea_level` dataFrame earlier, we didn't include the values from `global_sea_level` since the years were not aligned. Adding a column to a dataframe is as easy as adding an entry to a dictionary. So let's try:

In [ ]:
mean_sea_level["mean_global"] = global_sea_level["msl_ib_ns(mm)"]
mean_sea_level

The column is full of NaNs again because the auto-alignment feature of Pandas is searching for the index values like `1992.9323` in the index of `global_sea_level["msl_ib_ns(mm)"]` series and not finding them. Let's set its index to these years so that that auto-alignment can work for us and figure out which values we have and not:

In [ ]:
global_sea_level = global_sea_level.set_index("year")
global_sea_level["msl_ib_ns(mm)"]

In [ ]:
mean_sea_level["mean_global"] = global_sea_level["msl_ib_ns(mm)"]
mean_sea_level

**EXERCISE:** Create a new series containing the average of the 2 hemispheres minus the global value to see if that is close to 0. Work inside the mean_sea_level dataframe first. Then try with the original Series to see what happens with data alignment while doing computations.

In [ ]:
# Your code here

### Changing dtype of series

Now that the sea levels are looking pretty good, let's got back to the GISS temperature dataset. Because of the labels (strings) found in the middle of the timeseries, every column only assumed to contain strings (didn't convert them to floating point values):

In [ ]:
giss_temp.dtypes

That can be changed after the fact (and after the cleanup) with the `astype` method of a `Series`:

In [ ]:
giss_temp["Jan"].astype("float32")

In [ ]:
for col in giss_temp.columns:
    giss_temp.loc[:, col] = giss_temp[col].astype(np.float32)

An index has a `dtype` just like any Series and that can be changed after the fact too.

In [ ]:
giss_temp.index.dtype

For now, let's change it to an integer so that values can at least be compared properly. We will learn below to change it to a datetime object.

In [ ]:
giss_temp.index = giss_temp.index.astype(np.int32)

### Removing missing values

Removing missing values - once they have been converted to `np.nan` - is very easy. Entries that contain missing values can be removed (dropped), or filled with many strategies.

In [ ]:
full_globe_temp

In [ ]:
full_globe_temp.dropna()

In [ ]:
# This will remove any year that has a missing value. Use how='all' to keep partial years
giss_temp.dropna(how="any").tail()

In [ ]:
giss_temp.fillna(value=0).tail()

In [ ]:
# This fills them with the previous year. See also temp3.interpolate
giss_temp.fillna(method="ffill").tail()

Let's also mention the `.interpolate` method on a `Series`:

In [ ]:
giss_temp.Aug.interpolate().tail()

For now, we will leave the missing values in all our datasets, because it wouldn't be meaningful to fill them.

**EXERCISE:** Go back to the reading functions, and learn more about other options that could have allowed us to fold some of these pre-processing steps into the data loading.

## 4. Basic visualization

Now they have been formatted, visualizing your datasets is the next logical step and is trivial with Pandas. The first thing to try is to invoke the `.plot` to generate a basic visualization (uses matplotlib under the covers).

### Line plots

In [ ]:
full_globe_temp.plot()

In [ ]:
giss_temp.plot(figsize=LARGE_FIGSIZE)

In [ ]:
mean_sea_level.plot(subplots=True, figsize=(16, 12));

### Showing distributions information

In [ ]:
# Distributions of mean sean level globally and per hemisphere?
mean_sea_level.plot(kind="kde", figsize=(12, 8))

**QUIZ:** How to list the possible kinds of plots that the plot method can allow?

In [ ]:
# Distributions of temperature in each month since 1880
giss_temp.boxplot();

### Correlations

There are more plot options inside `pandas.tools.plotting`:

In [ ]:
# Is there correlations between the northern and southern sea level timeseries we loaded?
from pandas.tools.plotting import scatter_matrix
scatter_matrix(mean_sea_level, figsize=LARGE_FIGSIZE);

We will confirm the correlations we think we see further down...

## 5. Storing our work

For each `read_**` function to load data, there is a `to_**` method attached to Series and DataFrames.

**EXERCISE**: explore how the to_csv method work using `ipython`'s ? and store the `giss_temp` dataframe. Do the same to store the full_globe_temp series to another file.

Another file format that is commonly used is Excel, and there multiple datasets can be stored in 1 file.

In [ ]:
writer = pd.ExcelWriter("test.xls")

In [ ]:
giss_temp.to_excel(writer, sheet_name="GISS temp data")
full_globe_temp.to_excel(writer, sheet_name="NASA temp data")

In [ ]:
writer.close()

In [ ]:
with pd.ExcelWriter("test.xls") as writer:
    giss_temp.to_excel(writer, sheet_name="GISS temp data")
    pd.DataFrame({"Full Globe Temp": full_globe_temp}).to_excel(writer, sheet_name="FullGlobe temp data")

In [ ]:
%ls

Another, more powerful file format to store binary data, which allows us to store both `Series` and `DataFrame`s without having to cast anybody is HDF5.

In [ ]:
with pd.HDFStore("all_data.h5") as writer:
    giss_temp.to_hdf(writer, "/temperatures/giss")
    full_globe_temp.to_hdf(writer, "/temperatures/full_globe")
    mean_sea_level.to_hdf(writer, "/sea_level/mean_sea_level")
    local_sea_level_stations.to_hdf(writer, "/sea_level/stations")

**EXERCISE**: Add the greenhouse gas dataset in this data store. Store it in a separate folder.